0-module-and-module-source.emu

```diff
@@ -537,27 +569,46 @@ location: https://tc39.es/proposal-compartments/
+                1. If _handler_ is *null*, throw a *TypeError* exception.
+                1. If Type(_handler_) is Object, then
+                  1. Let _importHook_ be ? GetMethod(_handler_, *"importHook"*).
+                  1. If _importHook_ is *undefined*, then
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _moduleRecord_ and performs the following steps when called
+                      1. Assert: Type(_specifier_) is String.
+                      1. NOTE: Default import hook.
+                      1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
+                      1. Perform HostImportModuleDynamically(_moduleRecord_, _specifier_, _promiseCapability_).
+                      1. Return _promiseCapability_.[[Promise]].
+                  1. Else,
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _importHook_ and _handler_ and performs the following steps when called
+                      1. Assert: Type(_specifier_) is String.
+                      1. return ? Call(_importHook_, _handler_, « _specifier_ »).
+                  1. Let _importMetaHook_ be ? GetMethod(handler, *"importMetaHook"*).
+                  1. If _importMetaHook_ is *undefined*, then
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_importMeta_) that captures _moduleRecord_ and performs the following steps when called
+                      1. NOTE: Default import.meta hook.
+                      1. Let _importMetaValues_ be HostGetImportMetaProperties(_moduleRecord_).
+                      1. For each Record { [[Key]], [[Value]] } _p_ of _importMetaValues_, do
+                        1. Perform ! CreateDataPropertyOrThrow(_importMeta_, _p_.[[Key]], _p_.[[Value]]).
+                      1. Perform HostFinalizeImportMeta(_importMeta_, _moduleRecord_).
+                      1. Set _moduleRecord_.[[ImportMeta]] to _importMeta_.
+                      1. Return _importMeta_.
+                  1. Else,
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_importMeta_) that captures _importMetaHook_ and _handler_ and performs the following steps when called
+                      1. return ? Call(_importMetaHook_, _handler_, &laquo; _importMeta_ &raquo;).
+                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%Module.prototype%"*, « [[Module]], [[ModuleSourceInstance]], [[HooksThisValue]], [[ImportHook]], [[ImportMetaHook]] »).
                 1. Set _O_.[[ModuleSourceInstance]] to _moduleSource_.
                 1. Set _O_.[[ImportHook]] be CreateBuiltinFunction(_importHookClosure_, 1, *""*, &laquo; &raquo;).
+                1. Set _O_.[[ImportMetaHook]] be CreateBuiltinFunction(_importMetaHookClosure_, 0, *""*, &laquo; &raquo;).
+                1. Set _O_.[[HooksThisValue]] be _handler_.
```

Comment:
This is never used.

I slightly prefer to:
- keep `_O_.[[HooksThisValue]]`
- only store `_O_.[[ImportHook]] = _handler_.[[ImportHook]]` and `_O_.[[ImportMetaHook]] = _handler_.[[ImportMetaHook]]`
- call them as `Call(_O_.[[ImportHook]], _O_.[[HooksThisValue]], « _specifier_ »)`
- keep the fallback behavior where it's needed (in the runtime semantics of `import.meta`, or in the corresponding place for [[ImportHook]]), rather than in the module constructor

However, your current approach is also fine and we can delete `_O_.[[HooksThisValue]]`.

Reply:
ah, right... because they are part of the closures... got it! I think you're right on the fact that _O_.[[HooksThisValue]] can be deleted.

on the removal of the closures, let me see what I can do... I need to see if I have all the data at my disposal that the closure is not really needed. I was also thinking about hooks created by the spec rather than user-land... and trying to generalize that... but we will see.

Reply:
Ok, we are half way there, I think this is ok for now... we can refactor the importHook after we set on top of the 262 hook refactor @nicolo-ribaudo

Reply:
Ok, I have made the changes... now we hold the this value, and hooks could be undefined. The default behavior is implemented on the spot where those hooks are needed rather than during construction.